### DISCUSSION TITLE
Problems with charon

### DISCUSSION THREAD

**Discussion by xXConcasXx** [strongswan/strongswan#1210] [Q&A] [answered]
Hello, I'm having a problem with Strongswan, I cannot connect.
I am using a clean install of Ubuntu Server 22.04 LTS arm64 on Raspberry PI 3.

Running 

> tail -f /var/log/syslog

 gives me this output:

> Aug  9 20:39:37 raspberrypi charon: 00[DMN] Starting IKE charon daemon (strongSwan 5.9.5, Linux 5.15.0-1012-raspi, aarch64)
> Aug  9 20:39:37 raspberrypi charon: 00[LIB] providers loaded by OpenSSL: legacy default
> Aug  9 20:39:37 raspberrypi charon: 00[KNL] unable to create netlink socket: Protocol not supported (93)
> Aug  9 20:39:37 raspberrypi charon: 00[NET] installing IKE bypass policy failed
> Aug  9 20:39:37 raspberrypi charon: 00[NET] installing IKE bypass policy failed
> Aug  9 20:39:37 raspberrypi charon: 00[NET] enabling UDP decapsulation for IPv6 on port 4500 failed
> Aug  9 20:39:37 raspberrypi charon: 00[NET] installing IKE bypass policy failed
> Aug  9 20:39:37 raspberrypi charon: 00[NET] installing IKE bypass policy failed
> Aug  9 20:39:37 raspberrypi charon: 00[NET] enabling UDP decapsulation for IPv4 on port 4500 failed
> Aug  9 20:39:37 raspberrypi charon: 00[LIB] feature CUSTOM:libcharon in critical plugin 'charon' has unmet dependency: CUSTOM:kernel-ipsec
> Aug  9 20:39:37 raspberrypi charon: 00[NET] using forecast interface eth0
> Aug  9 20:39:37 raspberrypi charon: 00[CFG] joining forecast multicast groups: 224.0.0.1,224.0.0.22,224.0.0.251,224.0.0.252,239.255.255.250
> Aug  9 20:39:37 raspberrypi charon: 00[CFG] loading ca certificates from '/etc/ipsec.d/cacerts'
> Aug  9 20:39:37 raspberrypi charon: 00[CFG]   loaded ca certificate "C=US, O=Let's Encrypt, CN=R3" from '/etc/ipsec.d/cacerts/chain.pem'
> Aug  9 20:39:37 raspberrypi charon: 00[CFG] loading aa certificates from '/etc/ipsec.d/aacerts'
> Aug  9 20:39:37 raspberrypi charon: 00[CFG] loading ocsp signer certificates from '/etc/ipsec.d/ocspcerts'
> Aug  9 20:39:37 raspberrypi charon: 00[CFG] loading attribute certificates from '/etc/ipsec.d/acerts'
> Aug  9 20:39:37 raspberrypi charon: 00[CFG] loading crls from '/etc/ipsec.d/crls'
> Aug  9 20:39:37 raspberrypi charon: 00[CFG] loading secrets from '/etc/ipsec.secrets'
> Aug  9 20:39:40 raspberrypi charon: 00[CFG]   loaded RSA private key from '/etc/ipsec.d/private/privkey.pem'
> Aug  9 20:39:40 raspberrypi charon: 00[CFG]   loaded EAP secret for ggdigik
> Aug  9 20:39:40 raspberrypi charon: 00[CFG] loaded 0 RADIUS server configurations
> Aug  9 20:39:40 raspberrypi charon: 00[CFG] HA config misses local/remote address
> Aug  9 20:39:40 raspberrypi charon: 00[LIB] failed to load 1 critical plugin feature
> Aug  9 20:39:40 raspberrypi charon: 00[DMN] initialization failed - aborting charon

**Comment by strongX509** [comment]
* Is the IPsec stack and the XFRM interface loaded in the Linux kernel? (does the command `ip xfrm policy` work?)
* Does `charon` load the  `kernel-netlink` plugin?

**Comment by xXConcasXx** [comment]
Hello, thanks for your reply.
`ip xfrm policy` returns me this message:

> Cannot open netlink socket: Protocol not supported

I don't know how to see if `charon` is loading `kernel-netlink` plugin but I assume it isn't.

**Reply by strongX509** [reply]
This means that  IPsec or at least the xfrm kernel modules are not loaded in the kernel. See https://docs.strongswan.org/docs/5.9/install/kernelModules.html

**Reply by xXConcasXx** [reply]
This is currently my situation, how can I install the missing modules? I'm still a noob on the subject.

```
root@raspberrypi:/home/pi# grep '\<CONFIG_XFRM_USER\>' /boot/config-`uname -r`
CONFIG_XFRM_USER=m
root@raspberrypi:/home/pi# grep '\<CONFIG_NET_KEY\>' /boot/config-`uname -r`
CONFIG_NET_KEY=m
root@raspberrypi:/home/pi# grep '\<CONFIG_INET\>' /boot/config-`uname -r`
CONFIG_INET=y
root@raspberrypi:/home/pi# grep '\<CONFIG_IP_ADVANCED_ROUTER\>' /boot/config-`uname -r`
CONFIG_IP_ADVANCED_ROUTER=y
root@raspberrypi:/home/pi# grep '\<CONFIG_IP_MULTIPLE_TABLES\>' /boot/config-`uname -r`
CONFIG_IP_MULTIPLE_TABLES=y
root@raspberrypi:/home/pi# grep '\<CONFIG_INET_AH\>' /boot/config-`uname -r`
CONFIG_INET_AH=m
root@raspberrypi:/home/pi# grep '\<CONFIG_INET_ESP\>' /boot/config-`uname -r`
CONFIG_INET_ESP=m
root@raspberrypi:/home/pi# grep '\<CONFIG_INET_IPCOMP\>' /boot/config-`uname -r`
CONFIG_INET_IPCOMP=m
root@raspberrypi:/home/pi# grep '\<CONFIG_INET_XFRM_MODE_TRANSPORT\>' /boot/config-`uname -r`
root@raspberrypi:/home/pi# grep '\<CONFIG_INET_XFRM_MODE_TUNNEL\>' /boot/config-`uname -r`
root@raspberrypi:/home/pi# grep '\<CONFIG_INET_XFRM_MODE_BEET\>' /boot/config-`uname -r`
root@raspberrypi:/home/pi# grep '\<CONFIG_IPV6\>' /boot/config-`uname -r`
CONFIG_IPV6=y
root@raspberrypi:/home/pi# grep '\<CONFIG_INET6_AH\>' /boot/config-`uname -r`
CONFIG_INET6_AH=m
root@raspberrypi:/home/pi# grep '\<CONFIG_INET6_ESP\>' /boot/config-`uname -r`
CONFIG_INET6_ESP=m
root@raspberrypi:/home/pi# grep '\<CONFIG_INET6_IPCOMP\>' /boot/config-`uname -r`
CONFIG_INET6_IPCOMP=m
root@raspberrypi:/home/pi# grep '\<CONFIG_INET6_XFRM_MODE_TRANSPORT\>' /boot/config-`uname -r`
root@raspberrypi:/home/pi# grep '\<CONFIG_INET6_XFRM_MODE_TUNNEL\>' /boot/config-`uname -r`
root@raspberrypi:/home/pi# grep '\<CONFIG_INET6_XFRM_MODE_BEET\>' /boot/config-`uname -r`
root@raspberrypi:/home/pi# grep '\<CONFIG_IPV6_MULTIPLE_TABLES\>' /boot/config-`uname -r`
CONFIG_IPV6_MULTIPLE_TABLES=y
root@raspberrypi:/home/pi# grep '\<CONFIG_NETFILTER\>' /boot/config-`uname -r`
CONFIG_NETFILTER=y
root@raspberrypi:/home/pi# grep '\<CONFIG_NETFILTER_XTABLES\>' /boot/config-`uname -r`
CONFIG_NETFILTER_XTABLES=m
root@raspberrypi:/home/pi# grep '\<CONFIG_NETFILTER_XT_MATCH_POLICY\>' /boot/config-`uname -r`
CONFIG_NETFILTER_XT_MATCH_POLICY=m
root@raspberrypi:/home/pi# 
```

**Reply by Thermi** [reply]
install linux-modules and linux-modules-extra.

The latter contains xfrm_interface and xfrm_user.
Then you should have most if not all relevant packages.

**Reply by xXConcasXx** [reply]
I went to `https://packages.ubuntu.com/jammy/kernel/` and searched for my current version using `sudo uname -r`, in this case, `5.15.0-1005-raspi`, then I tried to install `linux-modules-5.15.0 -1005-raspi` but the terminal replied saying: 

> linux-modules-5.15.0-1005-raspi is already the newest version (5.15.0-1005.5).
> 0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.

then I tried to install `linux-modules-extra-5.15.0-1005-raspi`, doing this my kernel went from `5.15.0-1005-raspi` to `5.15.0-1013-raspi`, I rebooted and tried to check using 

> #! / bin / sh
> grep '\ <CONFIG_XFRM_USER \>' / boot / config-`uname -r`
> grep '\ <CONFIG_NET_KEY \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET \>' / boot / config-`uname -r`
> grep '\ <CONFIG_IP_ADVANCED_ROUTER \>' / boot / config-`uname -r`
> grep '\ <CONFIG_IP_MULTIPLE_TABLES \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET_AH \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET_ESP \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET_IPCOMP \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET_XFRM_MODE_TRANSPORT \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET_XFRM_MODE_TUNNEL \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET_XFRM_MODE_BEET \>' / boot / config-`uname -r`
> grep '\ <CONFIG_IPV6 \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET6_AH \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET6_ESP \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET6_IPCOMP \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET6_XFRM_MODE_TRANSPORT \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET6_XFRM_MODE_TUNNEL \>' / boot / config-`uname -r`
> grep '\ <CONFIG_INET6_XFRM_MODE_BEET \>' / boot / config-`uname -r`
> grep '\ <CONFIG_IPV6_MULTIPLE_TABLES \>' / boot / config-`uname -r`
> grep '\ <CONFIG_NETFILTER \>' / boot / config-`uname -r`
> grep '\ <CONFIG_NETFILTER_XTABLES \>' / boot / config-`uname -r`
> grep '\ <CONFIG_NETFILTER_XT_MATCH_POLICY \>' / boot / config-`uname -r

however I still notice the lack of xfrm modules except xfrm_user, do you have any advice for this situation?

**Reply by Thermi** [reply]
Do you not use apt and can install from it? Also, /boot/config* files are static, they do not change when you install any other package. They are contained in the kernel packages as far as I know.

It can be that a lot of modules were renamed in the meantime. I will need to check that, or you do that. You can just look for the relevant parts of the keywords and look for them, instead of relying on the whole config option name.

What is the result of the check script now that you got a newer kernel?

**Reply by xXConcasXx** [reply]
This is my output:
> root@raspberrypi:/home/pi# 
> grep '\<CONFIG_XFRM_USER\>' /boot/config-`uname -r`
> grep '\<CONFIG_NET_KEY\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET\>' /boot/config-`uname -r`
> grep '\<CONFIG_IP_ADVANCED_ROUTER\>' /boot/config-`uname -r`
> grep '\<CONFIG_IP_MULTIPLE_TABLES\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET_AH\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET_ESP\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET_IPCOMP\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET_XFRM_MODE_TRANSPORT\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET_XFRM_MODE_TUNNEL\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET_XFRM_MODE_BEET\>' /boot/config-`uname -r`
> grep '\<CONFIG_IPV6\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET6_AH\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET6_ESP\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET6_IPCOMP\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET6_XFRM_MODE_TRANSPORT\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET6_XFRM_MODE_TUNNEL\>' /boot/config-`uname -r`
> grep '\<CONFIG_INET6_XFRM_MODE_BEET\>' /boot/config-`uname -r`
> grep '\<CONFIG_IPV6_MULTIPLE_TABLES\>' /boot/config-`uname -r`
> grep '\<CONFIG_NETFILTER\>' /boot/config-`uname -r`
> grep '\<CONFIG_NETFILTER_XTABLES\>' /boot/config-`uname -r`
> grep '\<CONFIG_NETFILTER_XT_MATCH_POLICY\>' /boot/config-`uname -r`
> CONFIG_XFRM_USER=m
> CONFIG_NET_KEY=m
> CONFIG_INET=y
> CONFIG_IP_ADVANCED_ROUTER=y
> CONFIG_IP_MULTIPLE_TABLES=y
> CONFIG_INET_AH=m
> CONFIG_INET_ESP=m
> CONFIG_INET_IPCOMP=m
> CONFIG_IPV6=y
> CONFIG_INET6_AH=m
> CONFIG_INET6_ESP=m
> CONFIG_INET6_IPCOMP=m
> CONFIG_IPV6_MULTIPLE_TABLES=y
> CONFIG_NETFILTER=y
> CONFIG_NETFILTER_XTABLES=m
> CONFIG_NETFILTER_XT_MATCH_POLICY=m

Also I tried to do further Google searches and came across `libnl-xfrm-3-200` available at `https://packages.debian.org/` but after installation the result is still the same

**Comment by xXConcasXx** [comment, marked as answer]
I solved using Raspberry Pi OS instead of Ubuntu Server, I think Ubuntu Server variant for Raspberry is not suitable for hosting Strongswan.

Anyway thank you for your help, it was appreciated.

**Reply by tobiasbrunner** [reply]
I think this should have worked with Ubuntu too, it might just have required loading the modules manually if that's not happening automatically for some reason.

**Comment by fflrichard** [comment]
hi, i find we just need to do this and it will be ok
`apt install linux-modules-extra-raspi`

**Reply by webloginwu** [reply]
this solution sovle the problem
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.1 LTS Raspberry PI 4
Release:	22.04
Codename:	jammy
